src/zenml/orchestrators/input_utils.py

```diff
@@ -85,21 +85,30 @@ def resolve_step_inputs(
     for name, config_ in step.config.model_artifacts_or_metadata.items():
         issue_found = False
         try:
-            if config_.metadata_name is None and config_.artifact_name:
-                if artifact_ := config_.model.get_artifact(
+            context_model = config_._get_model_response(
```

Comment:
If I get this correctly, we assume here that:
- These model lazy loaders only work with the model of the pipeline context. So something like the following would fail:
```python
@pipeline
def my_pipeline():
  model = Model(name="my_model")
  lazy_artifact = model.get_artifact("my_artifact")
  my_step(lazy_artifact)
```
- The model must already be linked to the pipeline run, even if this is happening in the first step. Are we ensuring that already?

Reply:
Another thing that I thought of here: What happens in the case where steps define a separate model from the pipeline, in that case we don't link it to a run, right? In that case we would also run into an issue here?

For example:
```python
@step(model=Model("producer_model"))
def producer_step() -> Annotated[int, ArtifactConfig(model_name="pipeline_model")]:
  ...

@step(model=Model("consumer_model"))
def consumer_step(input):
  ...

@pipeline(model=Model("pipeline_model"))
def my_pipeline():
  producer_step()
  lazy_artifact = get_pipeline_context().model.get_artifact("my_artifact")
  consumer_step(lazy_artifact, after="producer_step")
```

Reply:
> ```python
> @step(model=Model("producer_model"))
> def producer_step() -> Annotated[int, ArtifactConfig(model_name="pipeline_model")]:
>   ...
> 
> @step(model=Model("consumer_model"))
> def consumer_step(input):
>   ...
> 
> @pipeline(model=Model("pipeline_model"))
> def my_pipeline():
>   producer_step()
>   lazy_artifact = get_pipeline_context().model.get_artifact("my_artifact")
>   consumer_step(lazy_artifact, after="producer_step")
> ```

You syntax is a bit off, but something like this will do:
```python
from typing import Annotated
from zenml import pipeline, step, Model, get_pipeline_context
from zenml.artifacts.artifact_config import ArtifactConfig


@step(model=Model(name="producer_model"))
def producer_step() -> Annotated[int, ArtifactConfig(name="my_artifact",model_name="pipeline_model",model_version="latest")]:
  return 1

@step(model=Model(name="consumer_model"))
def consumer_step(input:int):
  return input*2

@pipeline(model=Model(name="pipeline_model"), enable_cache=False)
def my_pipeline():
  producer_step()
  lazy_artifact = get_pipeline_context().model.get_artifact("my_artifact")
  consumer_step(lazy_artifact, after="producer_step")


if __name__ == "__main__":
    run = my_pipeline()
    print(run.model_version)
    print(run.config.model)
    ```

Reply:
> > If I get this correctly, we assume here that:
> > 
> > * These model lazy loaders only work with the model of the pipeline context. So something like the following would fail:
> > 
> > ```python
> > @pipeline
> > def my_pipeline():
> >   model = Model(name="my_model")
> >   lazy_artifact = model.get_artifact("my_artifact")
> >   my_step(lazy_artifact)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * The model must already be linked to the pipeline run, even if this is happening in the first step. Are we ensuring that already?
> 
> Yes, it will fail. But it is intended. As a side effect, I made some changes to how we warm-up models, to force pipeline model to be materialized earlier (before even first step `_prepare` is called).

Is this something that we should try to catch and fail, or do we just leave it as is?

Reply:
User gets this message in such cases: `ValueError: Failed to lazy load model version data in step ``consumer_step``: Lazy loading of the model failed, since the model version is not set in the pipeline context.`.
Shall it be more clear?

Reply:
Other option could be to check if the version is set in Model init, if it is called from the pipeline context and early raise. Maybe this is better, since this setting is inefficient anyway.

Reply:
I made changes which will show
```
ValidationError: 1 validation error for Model
  Value error, `version` must be set if you use the `Model` class directly in the pipeline body, otherwise, you can use `get_pipeline_context().model` to lazy load the current Model Version from the pipeline context. [type=value_error, 
input_value={'name': 'my_model'}, input_type=dict]
```

Reply:
Yes this early fail seems perfect, thanks!